kong/globalpatches.lua

```diff
@@ -15,6 +15,9 @@ return function(options)
   local meta = require "kong.meta"
 
 
+  require("cjson.safe").encode_sparse_array(nil, nil, 512)
```

Comment:
Two notes:

1. How about specifying `encode_sparse_array(false, 0)`? The first argument `false` doesn't has any signification when `ratio = 0`, but only to explicitly disable JSON Object conversion for maintainers/readers. Then `ratio = 0` in itself ensures that all sparse arrays (no matter their fill-in ratio) will be encoded as such. If we are concerned that this could become an attack vector by feeding cjson incredibly long and sparse arrays, we could enforce a hard limit on the total length of the array (i.e. last non-nil index). Maybe something like `2^15`, just to be safe. This depends on where this instance of cjson is used.
2. This setting most likely needs to be applied to `cjson` itself as well, and maybe (depending on the usage) to other instances of cjson too, possibly with different ratio settings.

The declarative config cjson instance could be:

```lua
local dc_cjson = cjson.new()
-- false: no JSON Object conversion
-- 0: no limit on how sparse the array may be
-- 2^15: max array index
dc_cjson.encode_sparse_array(false, 0, 2^15)
```

Reply:
Good advise @thibaultcha , I did some changes to make the limit 2^15 globally. On my machine, running the following case:

```shell
local cjson = require("cjson.safe")

cjson.encode_sparse_array(nil, nil, 2^15)

local a = {}

a[2^15] = 1

for i = 1, 1000 do
  local b = cjson.encode(a)
end

[datong.sun@localhost kong]$ time resty test.lua 

real	0m0.574s
user	0m0.561s
sys	0m0.005s
```

shows that serializing a sparse array as large as `2^15` should not affect the performance of the proxy by any meaningful amount to be an attractive attack vector.

We can't use `dc_cjson.encode_sparse_array(false, 0, 2^15)`, because the ratio of `0` disables the sparse array protection completely which means the `2^15` safe limit set later is never used. See the cJSON doc:

> Lua CJSON will never consider an array to be excessively sparse when ratio = 0.

Therefore, I think the risk of changing the default limit to `2^15` is minimum. Setting it separately for declarative config will be tricky because the use of PDK `response.exit` function, and is probably not necessary for the limit of `2^15`.